feat: server uploads (#214)
pingdotgg/uploadthing: docs/src/pages/api-reference/server.mdx

The UploadThing API Helper, for use ON YOUR SERVER. It's basically just a REST
API but better

### `uploadFiles`

`uploadFiles` let's you upload files directly from your server, without having
to define file routes. When uploadthing files from the server, no presigned URLs

> are used, your files are uploaded directly to S3 via our servers.

Reviewer: > your files are uploaded directly to S3 via our servers

- "directly" is not really correct here
- S3 bit is unnecessary, the backing storage doesn't matter to the end user. 

I think this could be changed to:

"your files are uploaded via our servers"

routes.

```tsx
import { utapi } from "uploadthing/server";

async function uploadFiles(formData: FormData) {
  "use server";
  const files = fd.getAll("files");
  const response = await utapi.uploadFiles(files);
  //    ^? { key: string, url: string }[]
}

function MyForm() {
  return (
    <form action={uploadFiles}>
      <input name="files" type="file" multiple />
      <button type="submit">Upload</button>
    </form>
  );
}
```

| Arguments | Type                       | Required | Description                       |
| :-------- | :------------------------- | :------- | :-------------------------------- |
| files     | FileEsque \|\| FileEsque[] | Yes      | The file(s) you want to upload    |
| metadata  | Json                       | No       | Metadata to attach to the file(s) |

```ts
// Edge / Node < 20 friendly File interface
interface FileEsque extends Blob {
  name: string;
}
```

### `uploadFileFromUrl`

`uploadFileFromUrl` is similar to `uploadFiles` as it let's you upload files
directly from your server, without having to define file routes.

When uploading files from URL, the file is first downloaded on **your** server,

> before being uploaded to S3 via our servers.

Reviewer: same as above.

The UploadThing API Helper, for use ON YOUR SERVER. It's basically just a REST
API but better

### `uploadFiles`

> Added in `v5.3`

`uploadFiles` let's you upload files directly from your server, without having

> to define file routes. When uploadthing files from the server, no presigned URLs

Reviewer: this is incorrect now.

This can be useful for uploading files from sever actions, or your own API
routes.

```tsx
import { utapi } from "uploadthing/server";

async function uploadFiles(formData: FormData) {
  "use server";
  const files = fd.getAll("files");
  const response = await utapi.uploadFiles(files);
  //    ^? { key: string, url: string }[]
}

function MyForm() {
  return (
    <form action={uploadFiles}>
      <input name="files" type="file" multiple />
      <button type="submit">Upload</button>
    </form>
  );
}
```

| Arguments | Type                       | Required | Description                       |
| :-------- | :------------------------- | :------- | :-------------------------------- |
| files     | FileEsque \|\| FileEsque[] | Yes      | The file(s) you want to upload    |
| metadata  | Json                       | No       | Metadata to attach to the file(s) |

```ts
// Edge / Node < 20 friendly File interface
interface FileEsque extends Blob {
  name: string;
}
```

### `uploadFilesFromUrl`

> Added in `v5.3`

`uploadFilesFromUrl` is similar to `uploadFiles` as it let's you upload files
directly from your server, without having to define file routes.

> When uploading files from URL, the file is first downloaded on **your** server,

Reviewer: may want to put this in a callout block -- since it could be costly for the user.

The UploadThing API Helper, for use ON YOUR SERVER. It's basically just a REST
API but better

### `uploadFiles`

> Added in `v5.3`

> `uploadFiles` let's you upload files without needing to define file routes. This
> can be useful for uploading files from sever actions, or your own API routes.

Reviewer 2: ```suggestion
Upload files directly from your server **without** using the file router. Useful for server-side file processing, uploading from a server action, and much more.
```

```tsx
import { utapi } from "uploadthing/server";

> async function uploadFiles(formData: FormData) {

Reviewer 2: Would be cool to have an example reading from FS, this is fine for now tho 🙏

  "use server";
  const files = fd.getAll("files");
  const response = await utapi.uploadFiles(files);
  //    ^? { key: string, url: string }[]
}

function MyForm() {
  return (
    <form action={uploadFiles}>
      <input name="files" type="file" multiple />
      <button type="submit">Upload</button>
    </form>
  );
}
```

When uploading files using `uploadFiles`, the files must be present on your
server. Then presigned URLs are generated on our servers before the files can be
uploaded to the storage provider.

| Arguments | Type                       | Required | Description                       |
| :-------- | :------------------------- | :------- | :-------------------------------- |
| files     | FileEsque \|\| FileEsque[] | Yes      | The file(s) you want to upload    |
| metadata  | Json                       | No       | Metadata to attach to the file(s) |

```ts
// Edge / Node < 20 friendly File interface
interface FileEsque extends Blob {
  name: string;
}
```

### `uploadFilesFromUrl`

> Added in `v5.3`

> `uploadFilesFromUrl` is similar to `uploadFiles` as it let's you upload files
> directly from your server, without having to define file routes.

Reviewer 2: ```suggestion
Have a file hosted somewhere else you want to upload on UploadThing? This is the function you're looking for
```

Reviewer 2: (also "lets" instead of "let's", generally the 's is for ownership)

<Callout type="warning">
  When uploading files from URL, the file is first downloaded on **your**
  server, before presigned URLs are created and the file is uploaded to the
  storage provider.
</Callout>

```tsx

> import { utapi } from "uploadthing/server";

Reviewer 2: IMO this example would be stronger if you made it simpler, i.e.

```tsx
const fileUrl = "https://test.com/some.png";

const uploadedFile = await utapi.uploadFileFromUrl(fileUrl);
```

> async function uploadFile(formData: FormData) {
>   "use server";
>   const url = fd.get("url");
>   const response = await utapi.uploadFileFromUrl(url);
>   //    ^? { key: string, url: string }
> }
> 
> function MyForm() {
>   return (
>     <form action={uploadFile}>
>       <input name="url" />
>       <button type="submit">Upload</button>
>     </form>
>   );
> }

Author: ```suggestion
const fileUrl = "https://test.com/some.png";

const uploadedFile = await utapi.uploadFileFromUrl(fileUrl);
```
